torch/distributions/constraints.py

```diff
@@ -456,16 +461,43 @@ def check(self, value):
         return _LowerCholesky().check(value) & unit_row_norm
 
 
-class _PositiveDefinite(Constraint):
+class _Square(Constraint):
     """
-    Constrain to positive-definite matrices.
+    Constrain to square matrices.
     """
     event_dim = 2
 
     def check(self, value):
-        # Assumes that the matrix or batch of matrices in value are symmetric
-        # info == 0 means no error, that is, it's SPD
-        return torch.linalg.cholesky_ex(value).info.eq(0).unsqueeze(0)
+        return torch.full(
+            size=value.shape[:-2],
+            fill_value=(value.shape[-2] == value.shape[-1]),
+            dtype=torch.bool,
+            device=value.device
+        )
+
+
+class _Symmetric(_Square):
+    """
+    Constrain to Symmetric square matrices.
+    """
+
+    def check(self, value):
+        square_check = super().check(value)
+        if not square_check.all():
+            return square_check
+        return value.eq(value.mT).all(-2).all(-1)
```

Comment:
I think the failures that you were seeing were due to precision errors. We can make it a softer check using, `torch.allclose(value, value.mT, atol=1e-6).all(-2).all(-1)`

Reply:
`torch.allclose` reduces all dimensions and there is no option to keep dimensions. I will find another solution.

Reply:
How about [isclose](https://pytorch.org/docs/stable/generated/torch.isclose.html)?

Reply:
I will try it out.